packages/react/src/Button/ButtonBase.tsx

```diff
@@ -65,46 +85,76 @@ const ButtonBase = forwardRef(
+          data-label-wrap={labelWrap}
+          aria-describedby={[loadingAnnouncementID, ariaDescribedBy]
+            .filter(descriptionID => Boolean(descriptionID))
+            .join(' ')}
+          // aria-labelledby is needed because the accessible name becomes unset when the button is in a loading state.
+          // We only set it when the button is in a loading state because it will supercede the aria-label when the screen
+          // reader announces the button name.
+          aria-labelledby={
+            loading ? [`${uuid}-label`, ariaLabelledBy].filter(labelID => Boolean(labelID)).join(' ') : ariaLabelledBy
+          }
+          id={id}
+          onClick={loading ? undefined : onClick}
+        >
+          {Icon ? (
+            loading ? (
+              <Spinner size="small" />
+            ) : (
+              <Icon />
+            )
+          ) : (
+            <>
+              <Box as="span" data-component="buttonContent" sx={getAlignContentSize(alignContent)}>
+                {loading && !LeadingVisual && !TrailingVisual && renderVisual(Spinner, loading, 'loadingSpinner')}
+                {LeadingVisual && renderVisual(LeadingVisual, Boolean(loading), 'leadingVisual')}
+                {children && (
+                  <span data-component="text" id={loading ? `${uuid}-label` : undefined}>
+                    {children}
+                    {count !== undefined && !TrailingVisual && (
+                      <CounterLabel data-component="ButtonCounter" sx={{ml: 2}}>
+                        {count}
+                      </CounterLabel>
+                    )}
+                  </span>
+                )}
+                {TrailingVisual && renderVisual(TrailingVisual, Boolean(loading) && !LeadingVisual, 'trailingVisual')}
+              </Box>
+              {TrailingAction && (
+                <Box as="span" data-component="trailingAction" sx={{...iconWrapStyles}}>
+                  <TrailingAction />
                 </Box>
               )}
-              {children && <span data-component="text">{children}</span>}
-              {count !== undefined && !TrailingVisual ? (
-                <Box as="span" data-component="trailingVisual" sx={{...iconWrapStyles}}>
-                  <CounterLabel data-component="ButtonCounter">{count}</CounterLabel>
-                </Box>
-              ) : TrailingVisual ? (
-                <Box as="span" data-component="trailingVisual" sx={{...iconWrapStyles}}>
-                  <TrailingVisual />
-                </Box>
-              ) : null}
-            </Box>
-            {TrailingAction && (
-              <Box as="span" data-component="trailingAction" sx={{...iconWrapStyles}}>
-                <TrailingAction />
-              </Box>
-            )}
-          </>
+            </>
+          )}
+        </StyledButton>
+        {loading && (
+          <VisuallyHidden>
+            <Status id={loadingAnnouncementID}>{loadingAnnouncement}</Status>
```

Comment:
Not a blocker, I think we can add this functionality after this PR has been merged!

We may want to add functionality that will only announce `loadingAnnouncement` once (n) seconds have passed. For example, if the loading state is toggled for more than 2 seconds, announce `loadingAnnouncement` through the live region by conditionally rendering the region after that amount of time has been passed.

Reply:
@joshblack is working on something to support that in https://github.com/primer/react/pull/4673

Going to wait until that merges, then update here. Thanks!